django/utils/deconstruct.py

```diff
@@ -39,6 +39,12 @@ def deconstruct(obj):
                     "For more information, see "
                     "https://docs.djangoproject.com/en/%s/topics/migrations/#serializing-values"
                     % (name, module_name, get_docs_version()))
+            if path and type(obj) is not klass:
+                return (
+                    '%s.%s' % (obj.__class__.__module__, name),
+                    obj._constructor_args[0],
+                    obj._constructor_args[1],
+                )
```

Comment:
I would move this check to the `return` below instead of doing a similar deconstruction twice:
```diff
diff --git a/django/utils/deconstruct.py b/django/utils/deconstruct.py
index aa952f6b73..9ba6bdd52f 100644
--- a/django/utils/deconstruct.py
+++ b/django/utils/deconstruct.py
@@ -40,7 +40,9 @@ def deconstructible(*args, path=None):
                     "https://docs.djangoproject.com/en/%s/topics/migrations/#serializing-values"
                     % (name, module_name, get_docs_version()))
             return (
-                path or '%s.%s' % (obj.__class__.__module__, name),
+                path
+                if path and type(obj) is klass
+                else '%s.%s' % (obj.__class__.__module__, name),
                 obj._constructor_args[0],
                 obj._constructor_args[1],
             )
```